bio/bio.go

```diff
@@ -0,0 +1,176 @@
+	Header() (DataTypeHeader, int64, error)
+	Next() (DataType, int64, error)
+}
+
+type Writer interface {
+	Write(io.Writer) (int, error)
+}
+
+/******************************************************************************
+
+Higher level parse
+
+******************************************************************************/
+
+type Parser[DataType fasta.Fasta, DataTypeHeader fasta.Header] struct {
+	LowLevelParser LowLevelParser[DataType, DataTypeHeader]
+}
+
+var emptyParser Parser[fasta.Fasta, fasta.Header] = Parser[fasta.Fasta, fasta.Header]{}
+
+func NewParser(format Format, r io.Reader) (*Parser[fasta.Fasta, fasta.Header], error) {
+	var maxLineLength int
+	switch format {
+	case Fasta:
+		maxLineLength = Fasta_DefaultMaxLineLength
+	}
+	return NewParserWithMaxLine(format, r, maxLineLength)
+}
+
+func NewParserWithMaxLine(format Format, r io.Reader, maxLineLength int) (*Parser[fasta.Fasta, fasta.Header], error) {
+	switch format {
+	case Fasta:
+		return &Parser[fasta.Fasta, fasta.Header]{LowLevelParser: fasta.NewParser(r, maxLineLength)}, nil
+	}
+	return nil, nil
+}
+
+/******************************************************************************
+
+Parser read functions
+
+******************************************************************************/
+
+func ReadWithMaxLine(format Format, path string, maxLineLength int) (*Parser[fasta.Fasta, fasta.Header], error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return &emptyParser, err
+	}
+	return NewParserWithMaxLine(format, file, maxLineLength)
+}
+
+func Read(format Format, path string) (*Parser[fasta.Fasta, fasta.Header], error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return &emptyParser, err
+	}
+	return NewParser(format, file)
+}
+
+/******************************************************************************
+
+Parser higher-level functions
+
+******************************************************************************/
+
+func (p *Parser[DataType, DataTypeHeader]) Next() (DataType, int64, error) {
+	return p.LowLevelParser.Next()
+}
+
+func (p *Parser[DataType, DataTypeHeader]) Header() (DataTypeHeader, int64, error) {
+	return p.LowLevelParser.Header()
+}
+
+func (p *Parser[DataType, DataTypeHeader]) ParseN(countN int) ([]DataType, error) {
+	return nil, nil
+}
+
+func (p *Parser[DataType, DataTypeHeader]) Parse() ([]DataType, error) {
+	return nil, nil
+}
+
+func (p *Parser[DataType, DataTypeHeader]) ParseAll() ([]DataType, DataTypeHeader, error) {
+	header, _, err := p.Header()
+	if err != nil {
+		return []DataType{}, DataTypeHeader{}, err
+	}
+	data, err := p.Parse()
+	if err != nil {
+		return []DataType{}, DataTypeHeader{}, err
+	}
+	return data, header, nil
+}
+
+func (p *Parser[DataType, DataTypeHeader]) ParseConcurrent(channel chan<- DataType) error {
+	return nil
+}
+
+/******************************************************************************
+
+Writer functions
```

Comment:
`io.MultiWriter()` has a type signature of `MultiWriter(writers ...Writer) Writer`, whereas `WriteAll` has a type signature of `WriteAll(data []io.WriterTo, header io.WriterTo, w io.Writer) error`. The writerTo are different, and generally they do different things: `WriteAll` writes many records + header to a single file, while `io.MultiWriter` splits one write to many writers, kinda like `tee`.

Reply:
`WriteFile()` is just legacy - pretty much all our previous packages have it. @TimothyStiles thoughts on including a `WriteFile()`? I'm leaning towards removing it, since os.Open is pretty simple.